### PR TITLE
improve printing of GroupedDataFrame in corner cases

### PR DESCRIPTION
**Pull request by bkamins**
Fixes https://github.com/JuliaData/DataFrames.jl/issues/3186

Covers all corner cases of various relative sizes of groups and single group.

### CODE REVIEW

File: src/groupeddataframe/show.jl [hunk 1]
```diff
@@ -42,30 +42,42 @@ function Base.show(io::IO, gd::GroupedDataFrame;
...
-            g1 = size(gd[1], 1) + 3
-            g2 = size(gd[N], 1) + 3
-
-            if g1 + g2 > h # won't fit on screen
-                if g1 < h ÷ 2
-                    h2 = h - g1 - 2 # show first group fully, squash last
-                elseif g2 < h ÷ 2
-                    h1 = h - g2 - 2 # show last group fully, squash first
+            # 2 lines for header and gap between groups, 3 lines for prompts;
+            # correcting for this allow at least 8 lines (4 for each group)
```
Comment (nalimilan) [inline src/groupeddataframe/show.jl:49]:
```suggestion
            # correcting for this allows at least 8 lines (4 for each group)
```

File: test/show.jl [hunk 2]
```diff
@@ -336,12 +336,24 @@ end
...
-    for h in 15:40
+    for a in 1:50, b in 1:50, h in 15:40
+        df = DataFrame(x = [fill(1, a); fill(2, b)])
         io = IOContext(IOBuffer(), :displaysize=>(h, 40), :limit=>true)
         show(io, groupby(df, :x), allcols=true)
         str = String(take!(io.io))
         nlines = length(split(str, '\n'))
-        desired = h - 3 # leave one line for last REPL prompt at top, two for new prompt
+        desired = min(a + b + 10, h - 3) # leave one line for last REPL prompt at top, two for new prompt
+        # (this is the same behavior as ungrouped data frames)
```
Comment (nalimilan) [inline test/show.jl:346]:
```suggestion
        # leave one line for last REPL prompt at top, two for new prompt
        # (this is the same behavior as ungrouped data frames)
        desired = min(a + b + 10, h - 3) 
```

File: test/show.jl [hunk 3]
```diff
@@ -336,12 +336,24 @@ end
...
+    end
+
+    for a in 1:50, h in 15:40
+        df = DataFrame(x = fill(1, a))
+        io = IOContext(IOBuffer(), :displaysize=>(h, 40), :limit=>true)
+        show(io, groupby(df, :x), allcols=true)
+        str = String(take!(io.io))
+        nlines = length(split(str, '\n'))
+        desired = min(a + 5, h - 3) # leave one line for last REPL prompt at top, two for new prompt
         # (this is the same behavior as ungrouped data frames)
```
Comment (nalimilan) [inline test/show.jl:357]:
```suggestion
        # leave one line for last REPL prompt at top, two for new prompt
        # (this is the same behavior as ungrouped data frames)
        desired = min(a + 5, h - 3)
```

**Review by nalimilan — APPROVED** [review]
I haven't checked the logic, but you know better...

**Comment by ronisbr** [comment]
LGTM!

**Review by ronisbr — APPROVED** [review]

**Comment by bkamins** [comment]
@nalimilan - it is good that we fixed these OutOfMemory issues on 32 bit machines as now our tests catch bugs instead of constantly failing.

@Jollywatt - if you are interested have a look at the last commit if you want to know how tests should be written so that they are correct on both 64 and 32 bit architectures.

File: src/groupeddataframe/show.jl [hunk 4]
```diff
@@ -42,30 +42,42 @@ function Base.show(io::IO, gd::GroupedDataFrame;
...
-                if g1 < h ÷ 2
-                    h2 = h - g1 - 2 # show first group fully, squash last
-                elseif g2 < h ÷ 2
-                    h1 = h - g2 - 2 # show last group fully, squash first
+            # 2 lines for header and gap between groups, 3 lines for prompts;
+            # correcting for this allows at least 8 lines (4 for each group)
+            h = max(h - 5, 8)
+
+            if N == 1
+                h1 = h + 3 # add two lines for prompts and one line as there no gap between groups
```
Comment (bkamins) [inline src/groupeddataframe/show.jl:53]:
```suggestion
                h1 = h + 3 # add two lines for prompts and one line as there is no gap between groups
```

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**